### ISSUE TITLE
new Ra Vulkan and D3D11 renderers show slightly different colors than OGL for many pixels

### ISSUE THREAD

**Issue by aufkrawall** [mpv-player/mpv#4979]
### mpv version and platform
mpv build priovided by rossy:
https://files.jrg.systems/mpv-d3d11-x86_64-20171012-git-e272dcc.7z
Windows 10 x64 AU

### Reproduction steps
Use a simple config like:
gpu-api=opengl
gpu-context=win
cscale=catmull_rom
scale=catmull_rom

### Expected behavior
gpu-api=d3d11 and vulkan should show the exact same color for every pixel like opengl.

### Actual behavior
Many pixels have slightly shifted colors with d3d11/vulkan when a shader based scaler is used.

So e.g. bilinear is not affected, but catmull_rom and ewa_lanczossharp are.
Both scale and cscale are affected. When shader based scalers are used for both scale and cscale at the same time, more pixels are different than with sole cscale.

OpenGL:
https://abload.de/img/oglw5sas.png

D3D11:
https://abload.de/img/d3d111yse0.png

Vulkan:
https://abload.de/img/vulkanmfspm.png

Different pixels between opengl and d3d11/vulkan marked:
https://abload.de/img/diff6ysnj.png

### Log file
Vulkan:
https://pastebin.com/EarUceDX

D3D11:
https://pastebin.com/NXM83rRp

OpenGL:
https://pastebin.com/aytY7Ui2

### Sample files
Used vid (careful, 2 frames):
http://www8.zippyshare.com/v/5TnwOaH8/file.html

I used this tool to mark the different pixels:
http://www19.zippyshare.com/v/LF1FmwHx/file.html

(that hoster works pretty well for me, at least with an adblocker)

**Comment by lachs0r** [comment]
Are you sure that’s not just dithering noise?

**Comment by garoto** [comment]
I was asking myself the same thing (but I tend to ask myself really dumb things in general).

**Comment by aufkrawall** [comment]
I only used the commands listed above, so dithering was disabled.
There are also 0 pixel difference between the d3d11 and vulkan screenshot, and there is also always 0 pixel difference when I create screenshots with same API several times (with mpv restarts in between).

**Comment by quilloss** [comment]
Could it be a difference in rounding?
Isn't Vulkan's rounding for floating point undefined?

```
ad2.mkv_opengl_ewa_lanczossharp.png - ad2.mkv_vulkan_ewa_lanczossharp.png
sum : 14
mean: 2.250514403292181e-06
max : 1
min : -1
= 0 : 6214290
< 0 : 3248
> 0 : 3262

ad2.mkv_opengl_catmull_rom.png - ad2.mkv_vulkan_catmull_rom.png
sum : 86619
mean: 0.01392409336419753
max : 1
min : -1
= 0 : 6002479
< 0 : 65851
> 0 : 152470
```

**Comment by aufkrawall** [comment]
It's at least not the FBO format, I checked with fbo-format=rgba32f and it confirms the other results.

**Comment by rossy** [comment]
Oh, I noticed there were rounding errors when testing --gpu-api=d3d11 against --gpu-api=opengl with [haasn/mpv-unit][1], but I didn't think of testing Vulkan as well. It's interesting that they're the same. I'm not sure how to explain that, though the pixel values only seem to only differ by ±1, so it's probably harmless. When I was testing with mpv-unit, I ended up comparing the images with ``compare -metric AE -fuzz 0.4%``, which ignores these ±1 differences.

Maybe the backends have defined rounding, but the way the textures are sampled is different when the image is rendered to the main framebuffer (with ra_fbo.flip = false.)

[1]: https://github.com/haasn/mpv-unit
